enarx-keep-sev-shim/asm/entry_ram64.S

```diff
@@ -0,0 +1,229 @@
+    mov    %cr0, %rax
+    and    $0x60050009, %eax # mask EMULATE_COPROCESSOR | MONITOR_COPROCESSOR
+    or     $0x80000021, %eax # set  PROTECTED_MODE_ENABLE | NUMERIC_ERROR | PAGING
+    mov    %rax, %cr0
+
+    # setup EFER
+    # EFER |= LONG_MODE_ACTIVE | LONG_MODE_ENABLE | NO_EXECUTE_ENABLE | SYSTEM_CALL_EXTENSIONS
+    # FIXME: what about already set bits?
+    mov    $0xc0000080, %ecx
+    rdmsr
+    or     $0xd01, %eax
+    mov    $0xc0000080, %ecx
+    wrmsr
+
+    # Setup the pagetables
+    # see also pagetables.s
+    # done dynamically, otherwise we would have to correct the dynamic symbols twice
+    lea  PML4T(%rip), %rax
+
+    lea  PDPT_OFFSET(%rip), %rbx
+    or   $0b11, %rbx    # (WRITABLE | PRESENT)
+    # store PDPT_OFFSET table in PML4T in the correct slot
+    # for KERNEL_OFFSET
+    mov  %rbx, (KERNEL_OFFSET >> 36)(%rax)
+
+    lea  PDPT_IDENT(%rip), %rbx
+
+    lea  PDT_IDENT(%rip), %rcx
+    or   $0b111, %rcx   # (USER_ACCESSIBLE | WRITABLE | PRESENT)
+    # store PDT_IDENT table in PDPT_IDENT in the correct slot
+    # 0x0 - 0x4000_0000
+    mov  %rcx, (%rbx)
+
+    or   $0b111, %rbx   # (USER_ACCESSIBLE | WRITABLE | PRESENT)
+    # store PDPT_IDENT table in PML4T in the correct slot
+    # 0x0 - ...
+    mov  %rbx, (%rax)
+
+    mov %rax, %cr3
+
+    # jump to RIP + KERNEL_OFFSET
+    lea _trampoline(%rip),%rax
+	mov $KERNEL_OFFSET, %rbx
+	adox %rbx, %rax
+	jmp *%rax
+_trampoline:
+
+    # load stack in kernel virtual address space
+    lea _initial_kernel_stack(%rip), %rsp
+    add $(INITIAL_STACK_SIZE-16), %rsp
+
+    # correct symbols with kernel virtual address space
+	mov $KERNEL_OFFSET, %r15
+	adox %r15, %rsi
+
+    # save rdi and add KERNEL_OFFSET
+    adox %rdi, %r15
+
+    # correct dynamic symbols with load offset + KERNEL_OFFSET
+	lea _DYNAMIC(%rip),%rdx
+
+    # the following is assembled from `rcrt1.c`
+
+	subq	$136, %rsp
+	xorl	%eax, %eax
+	movl	$32, %ecx
+	leaq	-120(%rsp), %r8
+	movq	%r8, %rdi
+	rep stosq
+	movq	(%rdx), %rcx
+	testq	%rcx, %rcx
+	je	.L2
+	leaq	8(%rdx), %rax
+	.p2align 4,,7
+	.p2align 3
+.L4:
+	cmpq	$31, %rcx
+	ja	.L3
+	movq	(%rax), %rdx
+	movq	%rdx, -120(%rsp,%rcx,8)
+.L3:
+	movq	8(%rax), %rcx
+	addq	$16, %rax
+	testq	%rcx, %rcx
+	jne	.L4
+.L2:
+	movq	16(%rsp), %rax
+	movq	24(%rsp), %rdx
+	addq	%rsi, %rax
+	testq	%rdx, %rdx
+	je	.L5
+	leaq	(%rax,%rdx), %rcx
+	.p2align 4,,7
+	.p2align 3
+.L7:
+	movq	8(%rax), %rdx
+	andl	$2147483647, %edx
+	cmpq	$8, %rdx
+	jne	.L6
+	movq	(%rax), %rdx
+	addq	%rsi, %rdx
+	addq	%rsi, (%rdx)
+.L6:
+	addq	$16, %rax
+	cmpq	%rcx, %rax
+	jne	.L7
+.L5:
+	movq	-64(%rsp), %rax
+	movq	-56(%rsp), %rdx
+	addq	%rsi, %rax
+	testq	%rdx, %rdx
+	je	.L1
+	leaq	(%rax,%rdx), %rcx
+	.p2align 4,,7
+	.p2align 3
+.L10:
+	movq	8(%rax), %rdx
+	andl	$2147483647, %edx
+	cmpq	$8, %rdx
+	jne	.L9
+	movq	16(%rax), %rdx
+	movq	(%rax), %rdi
+	addq	%rsi, %rdx
+	movq	%rdx, (%rdi,%rsi)
+.L9:
+	addq	$24, %rax
+	cmpq	%rcx, %rax
+	jne	.L10
+.L1:
+	addq	$136, %rsp
+
+    # end of `rcrt1.c`
```

Comment:
This section is neither auditable nor maintainable. I *think* that it is doing relocation. But I don't see why we can't do this in Rust.

Reply:
I might do this in "C" or Rust, but I think I need to correct the address of the function to call first with it... so chicken and egg.

Reply:
I don't think so. RIP-addressing shouldn't create a relocation entry. So as long as the function itself doesn't attempt to use an address, it should be fine. Therefore, you should be able to port the existing C code to Rust and call it directly.